statsmodels/base/distributed_estimation.py

```diff
@@ -0,0 +1,400 @@
+from statsmodels.base.elastic_net import fit_elasticnet
+from statsmodels.base.model import Results
+from statsmodels.tools.decorators import cache_readonly
+from statsmodels.regression.linear_model import OLS
+#from statsmodels.genmod.generalized_linear_model import GLM
+#from statsmodels.genmod.families import Binomial
+import statsmodels.base.wrapper as wrap
+import numpy as np
+
+"""
+distributed regularized estimation.
+
+Routines for fitting regression models using a distributed
+approach outlined in
+
+Jason D. Lee, Qiang Liu, Yuekai Sun and Jonathan E. Taylor.
+"Communication-Efficient Sparse Regression: A One-Shot Approach."
+arXiv:1503.04337. 2015. http://arxiv.org/abs/1503.04337.
+
+Currently the primary usecase is for out of memory computation,
+this approach currently does not support mutli-core processing
+and processing must be handled sequentially.
+"""
+
+
+def _gen_grad(mod, beta_hat, n, alpha, L1_wt, score_kwds):
+    """generates the log-likelihood gradient for the debiasing
+
+    Parameters
+    ----------
+    mod : statsmodels model class
+        The model for the current machine.
+    beta_hat : array-like
+        The estimated coefficients for the current machine.
+    n : scalar
+        machine specific sample size.
+    alpha : scalar or array-like
+        The penalty weight.  If a scalar, the same penalty weight
+        applies to all variables in the model.  If a vector, it
+        must have the same length as `params`, and contains a
+        penalty weight for each coefficient.
+    L1_wt : scalar
+        The fraction of the penalty given to the L1 penalty term.
+        Must be between 0 and 1 (inclusive).  If 0, the fit is
+        a ridge fit, if 1 it is a lasso fit.
+    score_kwds : dict-like or None
+        Keyword arguments for the score function.
+
+    Returns
+    -------
+    An array-like object of the same dimension as beta_hat
+
+    Notes
+    -----
+    In general:
+
+    nabla l_k(beta)
+
+    For the simple linear case:
+
+    X^T(y - X^T beta) / n
+    """
+
+    grad = -mod.score(np.r_[beta_hat], **score_kwds) / n
+    # this second part comes from the elastic net penalty
+    grad += alpha * (1 - L1_wt)
+    return grad
+
+
+def _gen_wdesign_mat(mod, beta_hat, hess_kwds):
+    """generates the weighted design matrix necessary to generate
+    the approximate inverse covariance matrix
+
+    Parameters
+    ----------
+    mod : statsmodels model class
+        The model for the current machine.
+    beta_hat : array-like
+        The estimated coefficients for the current machine.
+    hess_kwds : dict-like or None
+        Keyword arguments for the hessian function.
+
+    Returns
+    -------
+    An array-like object, updated design matrix, same dimension
+    as mod.exog
+    """
+
+    # TODO need to handle duration and other linear model classes
+    return np.sqrt(mod.hessian_obs(np.r_[beta_hat], **hess_kwds)).dot(mod.exog)
```

Comment:
If you get `hessian_obs` to return only the diagonal vector, something like this should work:

```
hess = mod.hessian_obs(beta_hat, **hess_kwds)
rhess = np.sqrt(hess)
return rhess[:, None] * mod.exog
```

Also, I'm not sure why you need the `np.r_[beta_hat]`, you might not need it, or just use `np.asarray(beta_hat)`.